photon-ml/src/main/scala/com/linkedin/photon/ml/model/RandomEffectModel.scala

```diff
@@ -140,29 +140,34 @@ object RandomEffectModel {
       randomEffectType: String,
       featureShardId: String): KeyValueScore = {
 
+    val hashPartitioner = new HashPartitioner(dataPoints.getNumPartitions)
+
+    /*
+     * We perform a replicated partitioned hash join here under the assumption that we can fit the per partition
+     * random effect models in memory. We first partition both relations using the same partitioner and then zip them.
+     * This ensures that the same keys from both relations go in the same partition. Given above, we can now perform the
+     * join by doing the following operations per partition:
+     *   1. Load the random effect models in memory
+     *   2. Iterate over the data points
+     *   3. For each data point, look up the corresponding random effect model in the in memory map and score
+     */
     val scores = dataPoints
-      .map { case (uniqueId, gameData) =>
-        val randomEffectId = gameData.idTypeToValueMap(randomEffectType)
-        val features = gameData.featureShardContainer(featureShardId)
-        (randomEffectId, (uniqueId, features))
+      .map { case (uniqueId, gameDatum) =>
+        val randomEffectId = gameDatum.idTypeToValueMap(randomEffectType)
+        (randomEffectId, (uniqueId, gameDatum))
       }
-      .cogroup(modelsRDD)
-      .flatMap { case (randomEffectId, (uniqueIdAndFeaturesIterable, modelsIterable)) =>
-        // TODO(fastier): we should move that precondition upfront and check it only once, for speed
-        assert(modelsIterable.size <= 1,
-          s"More than one model (${modelsIterable.size}) found for individual Id $randomEffectId of " +
-            s"random effect type $randomEffectType")
-
-        if (modelsIterable.isEmpty) {
-          uniqueIdAndFeaturesIterable.map { case (uniqueId, _) => (uniqueId, 0.0) }
-        } else {
-          val model = modelsIterable.head
-          uniqueIdAndFeaturesIterable.map { case (uniqueId, features) =>
-            (uniqueId, model.computeScore(features))
+      .partitionBy(hashPartitioner)
+      .zipPartitions(modelsRDD.partitionBy(hashPartitioner))(
+        (dataIt, modelIt) => {
+          val lookupTable = modelIt.toMap
+          dataIt.map {
```

Comment:
I like it! Only materializing the smaller object, which is the model, is very smart.